[DataPipe] Adding UnBatcherMapDataPipe

NivekT opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68197
* __->__ #68196

Implements part of #57031

cc @VitalyFedyunin @ejguan @NivekT

Differential Revision: [D32440965](https://our.internmc.facebook.com/intern/diff/D32440965)

NivekT committed:
[DataPipe] Adding UnBatcherMapDataPipe

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding UnBatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding UnBatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT committed:
Update on "[DataPipe] Adding UnBatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440965).

NivekT commented on torch/utils/data/datapipes/map/grouping.py:
```diff
@@ -0,0 +1,89 @@
+            pass
+        return level
+
+    def __getitem__(self, index) -> T:
```
The complicated logic is here. This part maps the new, given index to the elements within input, batched DataPipe.
The logic assumes that the input DataPipe is batched properly with full batches (except maybe the last batch), otherwise, an exception is raised.

NivekT committed:
Update on "[DataPipe] Adding UnBatcherMapDataPipe"




cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440965](https://our.internmc.facebook.com/intern/diff/D32440965)

[ghstack-poisoned]

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440965).

ejguan commented on torch/utils/data/datapipes/map/grouping.py:
```diff
@@ -0,0 +1,89 @@
+
+    def _get_batch_sizes(self) -> List[int]:
+        sizes = [len(self.datapipe)]
+        curr = self.datapipe.__getitem__(0)
```
Please use `[0]` rather than `__getitem__(0)`.

ejguan commented on torch/utils/data/datapipes/map/grouping.py:
```diff
@@ -0,0 +1,89 @@
+
+    def __getitem__(self, index) -> T:
+        batch_sizes: List[int] = self._get_batch_sizes()
+        if self.unbatch_level == 0:
+            return self.datapipe[index]  # type: ignore[return-value]
```
Is there any change `unbatch_level` is going to be 0?

ejguan reviewed (commented):
This is way too complicated. I would load whole dataset (prior datapipe) into memory for this operation if needed to get rid of all these crazy logics with a warning.

NivekT commented:
> This is way too complicated. I would load whole dataset (prior datapipe) into memory for this operation if needed to get rid of all these crazy logics with a warning.

As in loading the whole dataset into memory and unbatching once up front? That will potentially have out of memory issue, correct?

ejguan commented:
> As in loading the whole dataset into memory and unbatching once up front? That will potentially have out of memory issue, correct?

Yeah. But, in current implementation, we have a dangerous assumption that each batch from prior datapipe would have exact same shape, which is not always true.

NivekT committed:
Update on "[DataPipe] Adding UnBatcherMapDataPipe"


Implements part of #57031

cc @VitalyFedyunin ejguan @NivekT

Differential Revision: [D32440965](https://our.internmc.facebook.com/intern/diff/D32440965)

[ghstack-poisoned]

NivekT commented:
> > As in loading the whole dataset into memory and unbatching once up front? That will potentially have out of memory issue, correct?
> 
> Yeah. But, in current implementation, we have a dangerous assumption that each batch from prior datapipe would have exact same shape, which is not always true.

More specifically - at each level, each batch except for the last has the same size. That is the behavior that you get if you create the batches via `Batcher`. I think that is an acceptable assumption? Do you know if users usually do batching outside of batcher?

This test case shows that batches like this still work:
https://github.com/pytorch/pytorch/blob/9472340a6022ccf8496b2f75a8e9330cf8f67072/test/test_datapipe.py#L1662-L1668

Are there some specific situations you have in mind that will not work this implementation?

NivekT commented:
@nivekt has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32440965).

ejguan commented:
> > > As in loading the whole dataset into memory and unbatching once up front? That will potentially have out of memory issue, correct?
> > 
> > 
> > Yeah. But, in current implementation, we have a dangerous assumption that each batch from prior datapipe would have exact same shape, which is not always true.
> 
> More specifically - at each level, each batch except for the last has the same size. That is the behavior that you get if you create the batches via `Batcher`. I think that is an acceptable assumption? Do you know if users usually do batching outside of batcher?
> 
> This test case shows that batches like this still work:
> 
> https://github.com/pytorch/pytorch/blob/9472340a6022ccf8496b2f75a8e9330cf8f67072/test/test_datapipe.py#L1662-L1668
> 
> Are there some specific situations you have in mind that will not work this implementation?

I think for text or audio that each batch can have different size like `[[1,2], [2,3,4], [5,6,7,8]]`

NivekT commented:
> [[1,2], [2,3,4], [5,6,7,8]]

@ejguan With an input like that, the current implementation will return the first two elements of each batch because the first batch has length 2. That is not what we want for sure.

```
arr = [[1,2], [2,3,4], [5,6,7,8]]
input_dp = dp.map.SequenceWrapper(arr)
unbatch_dp: MapDataPipe = dp.map.UnBatcher(input_dp)
print(list(unbatch_dp))
> [1, 2, 2, 3, 5, 6]
```

Still, I do wonder if the current implementation cover sufficient cases (>80-90%?) of cases such that we should keep it.

@VitalyFedyunin please let us know if you have any thoughts as well.

ejguan commented:
> Still, I do wonder if the current implementation cover sufficient cases (>80-90%?) of cases such that we should keep it.

I think the main problem here is the inconsistent behavior with `UnbatcherIterDataPipe`. It can cause confusion for users.

NivekT commented:
> > Still, I do wonder if the current implementation cover sufficient cases (>80-90%?) of cases such that we should keep it.
> 
> I think the main problem here is the inconsistent behavior with `UnbatcherIterDataPipe`. It can cause confusion for users.

Yep, I think if we do keep it, it probably should have a different name (maybe `unbatch_lazily`) or have an input option to determine the behavior.

NivekT commented:
Closing for now since this feature may not be wanted/needed
